code-scan.js

```diff
@@ -0,0 +1,203 @@
+  });
+}
+
+async function octokitRequest(request) {
+  console.log(`Running ${request} function`);
+  try {
+    // few functions require different properties
+    let requestProperties = { owner, repo };
+    switch (request) {
+      case "createFork":
+        requestProperties.owner = originalOwner;
+        requestProperties.organization = owner;
+        break;
+    }
+    const response = await octokitFunctions[request](requestProperties);
+    console.log(`Function ${request} finished succesfully`);
+    return response.data;
+  } catch (error) {
+    console.log(`Failed to run ${request}: ${error.message}`);
+  }
+}
+
+async function putRequest(request) {
+  //generic function for PUT requests
+  try {
+    await octokit.request(`PUT /repos/{owner}/{repo}/${request}`, {
+      owner,
+      repo,
+    });
+  } catch (error) {
+    console.log(`Failed to run ${request}: ${error.message}`);
+  }
+}
+
+async function getSha(ref) {
+  response = await octokit.rest.git.getRef({
+    owner,
+    repo,
+    ref: `heads/${ref}`,
+  });
+  return response.data;
+}
+
+async function deleteExistingWorkflows(sha){
+  console.log(`Delete existing workflows`)
+  await octokit.rest.repos.deleteFile({
+    owner,
+    repo,
+    path: ".github/workflows/codeql-analysis.yml",
+    message: "🤖 Delete existing workflows",
+    sha,
+  });
+
+}
+
+async function pushWorkflowFile() {
+  let languages = await octokitRequest("listLanguages");
+  languages = `${JSON.stringify(Object.keys(languages))}`;
+  console.log(`Detected languages: ${languages}`);
+
+  console.log(`Add Codeql workflow file`);
+  let workflowFile = fs.readFileSync("codeql-analysis-check.yml", "utf8");
+  workflowFile = workflowFile.replace("languageString", languages);
+
+  console.log(`Add Codeql workflow file`);
+
+  try {
+    const response = await octokit.request(
+      "PUT /repos/{owner}/{repo}/contents/.github/workflows/codeql-analysis-check.yml",
+      {
+        owner,
+        repo,
+        path: ".github/workflows/check-and-validate-codeql.yml",
+        message: "Inject codeql workflow",
+        content: Buffer.from(workflowFile).toString("base64"),
+      }
+    );
+
+    console.log("Workflow file created successfully");
+  } catch (error) {
+    console.error("Error creating workflow file:", error);
+  }
+}
+
+async function triggerCodeqlScan(workflow_id, ref) {
+  console.log(`Trigger codeql scan`);
+  await octokit.rest.actions.createWorkflowDispatch({
+    owner,
+    repo,
+    workflow_id,
+    ref,
+  });
+}
+
+async function waitForCodeqlScan() {
+  console.log(`Get the dispatched run id`);
+  const response = await octokit.rest.actions.listWorkflowRunsForRepo({
+    owner,
+    repo,
+    event: "workflow_dispatch",
+  });
+
+  const run_id = response.data.workflow_runs[0].id;
+
+  let status = "queued";
+  while (status != "completed") {
+    console.log(`Wait for scan to complete - Run id : ${run_id}`);
+    await wait(15000);
```

Comment:
This scan can potentially take a minute or more. Checking should be done in a loop until successful.

Reply:
nvm, it's already a look 🤦